CXSPA-574: Customer ID Http Header for ASM emulation

justinlee01 opened the pull request:
Closes https://jira.tools.sap/browse/CXSPA-574 , https://jira.tools.sap/browse/CXSPA-149 , https://jira.tools.sap/browse/CXSPA-151

- Creating UserIdInterceptor, used to append a "sap-commerce-cloud-user-id" header to certain HTTP requests that provide a HttpContext. This is so that there is no ambiguity as to who (the agent or the emulated user) is fetching certain resources.
- Covering GET /costcenters and GET /product/search
- Changing some E2E utils to be more stable (ex. use in-house `interceptGet` rather than cy.get with an incorrect BASE_SITE)
- Changing some E2E utils to support the B2B case

justinlee01 committed:
Creating UserIdInterceptor, that appends sap-commerce-cloud-user-id header to requests for certain endpoints when a customer is emulated

justinlee01 committed:
E2E tests for ASM apparel site

justinlee01 committed:
UserIdInterceptor can be configured with Injectable for its allowlist of paths. Injectable made for OCC user ID constants.

justinlee01 committed:
The response type for requests in UserIdInterceptor is irrelevant, so changing any to unknown for precision

justinlee01 committed:
Using GlobService in UserIdInterceptor to match path patterns

justinlee01 committed:
Moving UserIdInterceptor to ASM module. UserIdInterceptor only gets constructed when the ASM module is imported.
UserIdInterceptor pattern configuration is in the ASM module's configuration.
OCC user IDs are injected in a stable way.

justinlee01 committed:
Deleting obsolete config file

justinlee01 committed:
Adding ASM B2B test using costcenters. When listening for user details in B2B, listen for /orgUsers/ endpoint. Using internal interceptGet for addProductToCart helper

justinlee01 committed:
No longer using glob patterns to identify which endpoints need the sap-commerce-cloud-user-id header - instead, using HttpContext for certain endpoints; the UserIdInterceptor looks for a specific context to know which endpoints to send the header with.

justinlee01 committed:
Fixing unit tests for UserIdInterceptor

justinlee01 committed:
Adding empty spec file to remind myself to finish chores when a PR is made for this branch

justinlee01 committed:
UserIdInterceptor looks for a flag in ASM context before sending the user ID header

justinlee01 committed:
Unit test for ASMContextService

justinlee01 committed:
Merge branch 'develop' into feature/CXSPA-151

justinlee01 committed:
Adding copyright header to new files

plabadie commented on feature-libs/asm/root/interceptors/user-id.interceptor.ts:
```diff
@@ -0,0 +1,76 @@
+      userIdObservable = of(asmContext.sendUserIdAsHeader);
+    } else if (asmContext.sendUserIdAsHeader) {
+      userIdObservable = this.userIdService
+        .getUserId()
```
consider calling `takeUserId()` instead of `getUserId()` if we just need the current userId and move on.

justinlee01 replied:
! I didn't realize the difference. Thank you.

plabadie commented on projects/core/src/occ/services/asm-context.service.ts:
```diff
@@ -0,0 +1,26 @@
+@Injectable({
+  providedIn: 'root',
+})
+export class AsmContextService {
```
For the sake of the discussion, would there be a way to make this context mechanism not ASM specific?  Ideally we'd avoid bringing back ASM logic back in the core lib and spreading ASM related dependencies in classes.

justinlee01 replied:
Hmm. I can rename the method `createASMContext` (same implementation) and thus rename the service to `HttpContextService`.

armanozak replied:
Or completely remove it maybe? After all, it doesn’t do much and `HttpContext` is chainable. It’s the same as doing `new HttpContext().set(OCC_ASM_TOKEN, contextObj)` and the `OCC_ASM_TOKEN` is already exposed.

plabadie replied:
I agree that we should rather remove it.

plabadie commented on feature-libs/asm/root/interceptors/user-id.interceptor.ts:
```diff
@@ -0,0 +1,76 @@
+ * configure a request with 'sap-commerce-cloud-user-id' header.
+ */
+@Injectable({ providedIn: 'root' })
+export class UserIdInterceptor implements HttpInterceptor {
```
Could this logic become non ASM specific and make it's way in the core lib?  Let's discuss.

justinlee01 replied:
That's an interesting discussion. It was moved to the ASM module based on this premise:
1. 'sap-commerce-cloud-user-id' only needs to be sent when the user needs to be explicit i.e. in customer emulation case.
2. If it's only for customer emulation, then we can bolster the code's performance, if ever so slightly, by removing this interceptor from the chain, and by needing to load the code in the first place.

Is there an incentive to move it to core?

plabadie replied:
In OCC/backend, is this header used exclusively for ASM?

justinlee01 replied:
The header essentially overrides any session context that may have been used. For example, passing in the emulated customer ID will always override the user agent's authorization with the emulated user's. It's a very straightforward, if blunt, tool.

I can only imagine this would be used for ASM _for now_. I can't imagine any other scenario, at least given the current use of SPA.

plabadie replied:
I'm ok to leave the interceptor in asm.  I agree that for the foreseeable future, the purpose of the interceptor is to enable customer emulation for endpoints that don't have other ways to pass the user id parameter.

I think however, that the logic of assigning a context and the required objects in "core" should be named in a more generic, non asm way.  We never know what future use case we might have for it or how customers might use it.  Plus this keeps the core lib free of apparent asm specific logic.  I left a comment in the file where the token and interface are declared for the context.

plabadie commented on feature-libs/asm/root/interceptors/user-id.interceptor.ts:
```diff
@@ -0,0 +1,76 @@
+      })
+    );
+  }
+}
```
Overall we might investigate to see if in some cases, it's best to bail out of the interceptor logic.  For example, perhaps it doesn't make sense to pass `anonymous` or `current` as a customer id in the header.  Figuring out that rule might require a bit of investigation.

justinlee01 replied:
Line 55 serves that purpose -
```
this.uniqueUserIdConstants.has(userId) ? undefined : userId
```
This filters out any hardcoded userId (that is not a real UUID) as that will break the request, as I've found out.

justinlee01 committed:
Changing getUserId() to takeUserId(); did not realize we were listening to a hot stream

plabadie commented on projects/core/src/occ/utils/index.ts:
```diff
@@ -4,6 +4,17 @@
+
+export const OCC_USER_ID_CONSTANTS = new InjectionToken<{
+  [identifier: string]: string;
+}>('List of OCC constants that pass for user IDs.');
```
Couldn't we let other classes import the constants instead of requiring dependency injection for them to use the constants?

plabadie commented on projects/core/src/occ/base-occ.module.ts:
```diff
@@ -27,6 +29,10 @@ export class BaseOccModule {
+        {
+          provide: OCC_USER_ID_CONSTANTS,
+          useValue: OCC_USER_IDS,
+        },
```
Another option would be to provide the token as we instantiate it like so:
```
export const OCC_USER_ID_CONSTANTS = new InjectionToken('OCC_USER_ID_CONSTANTS', {
  providedIn: 'root',
  factory() {
    return OCC_USER_IDS;
  }
});
```

Got this idea from @armanozak and I thought it might be more concise to go with this.  What do you think?

justinlee01 replied:
Sure, I'll implement that. I didn't even know it accepted a second argument.

justinlee01 committed:
Value for OCC_ASM_USER_ID_CONSTANTS is provided in the InjectionToken itself. Getting rid of AsmContextService and leaving a comment on each line.

justinlee01 committed:
Merge branch 'develop' into feature/CXSPA-574

plabadie commented on projects/storefrontapp-e2e-cypress/cypress/helpers/asm.ts:
```diff
@@ -37,8 +37,12 @@ export function listenForCustomerSearchRequest(): string {
+    return interceptGet('userDetails', '/users/*');
+  } else {
+    return interceptGet('userDetails', '/orgUsers/*');
+  }
```
Suggestion to improve readability, do :
```
if (b2b) {
     ...
} else {
     ...
}
```

plabadie commented on projects/core/src/occ/utils/occ-asm-token.ts:
```diff
@@ -0,0 +1,19 @@
+  sendUserIdAsHeader?: string | boolean;
+}
+
+export const OCC_ASM_TOKEN = new HttpContextToken<OccAsmContext>(() => ({}));
```
I think it would help readability to call the token `OCC_ASM_CONTEXT_TOKEN` or even `OCC_ASM_HTTP_CONTEXT_TOKEN`.

Or even something more generic than ASM as mentioned above.

plabadie commented on projects/core/src/occ/adapters/user/occ-user-cost-centers.adapter.ts:
```diff
@@ -25,8 +26,15 @@ export class OccUserCostCenterAdapter implements UserCostCenterAdapter {
+    const context = new HttpContext().set(OCC_ASM_TOKEN, {
+      sendUserIdAsHeader: true,
+    });
+
```
Is there unit test coverage we could add for this?  (depending on how complex that would be, I don't know what it requires)

plabadie commented on projects/core/src/occ/adapters/product/occ-product-search.adapter.ts:
```diff
@@ -38,8 +39,13 @@ export class OccProductSearchAdapter implements ProductSearchAdapter {
+    // @see feature-libs/asm/root/interceptors/user-id.interceptor.ts
+    const context = new HttpContext().set(OCC_ASM_TOKEN, {
+      sendUserIdAsHeader: true,
+    });
```
Is there unit test coverage we could add for this?  (depending on how complex that would be, I don't know what it requires)

plabadie commented on projects/core/src/occ/utils/occ-asm-token.ts:
```diff
@@ -0,0 +1,19 @@
+
+import { HttpContextToken } from '@angular/common/http';
+
+export interface OccAsmContext {
```
Perhaps the "core lib" side of the logic could be called something else than "Asm" specific names and make it more generic.  It's really just about sending a flag or the user id to ask for the user id to be added as a header.

I don't have specific names in mind, but it could be something like `UserIdHttpContext`.

Also, while we are discussing naming, the file could be named `occ-userid-http-context.ts` or something like that as long as it it consistent with the naming inside the file.

justinlee01 replied:
The original idea was that this object could be extended for other flags needed for customer emulation. However, there is currently nothing else to send.
I am inclining toward your suggestion. Any thoughts @armanozak ?

armanozak replied:
How about calling the interface just `OccHttpContext` and the token `OCC_HTTP_CONTEXT`? Maybe later we can even consider making the context interface augmentable.

plabadie commented on feature-libs/asm/root/interceptors/user-id.interceptor.ts:
```diff
@@ -0,0 +1,76 @@
+ */
+@Injectable({ providedIn: 'root' })
+export class UserIdInterceptor implements HttpInterceptor {
+  private readonly userIdHeader = 'sap-commerce-cloud-user-id';
```
Can we change the private members to `protected`?

`protected` is the default visibility we use over private.  This helps for extensibility (customers can extend the class and see/use the variables.)

plabadie commented on feature-libs/asm/root/interceptors/user-id.interceptor.ts:
```diff
@@ -0,0 +1,76 @@
+  constructor(
+    protected userIdService: UserIdService,
+    @Inject(OCC_USER_ID_CONSTANTS)
+    userIdConstants: { [identifier: string]: string }
```
Just for consistency, could the userIdConstants property be protected as well?

plabadie commented on feature-libs/asm/root/interceptors/user-id.interceptor.ts:
```diff
@@ -0,0 +1,76 @@
+ * configure a request with 'sap-commerce-cloud-user-id' header.
+ */
+@Injectable({ providedIn: 'root' })
+export class UserIdInterceptor implements HttpInterceptor {
```
I think that precise naming improves the code readability, may I suggest we give the interceptor a name that communicates its purpose.  Perhaps `UserIdHeaderInterceptor` or `UserIdHttpHeaderInterceptor`

justinlee01 committed:
Private properties in UserIdInterceptor are now protected, for extensibility. Switching code in E2E tests for readability.

justinlee01 committed:
Renaming UserIdInterceptor -> UserIdHttpHeaderInterceptor. Also adding expectations to unit tests for modified adapters.

justinlee01 committed:
Merge branch 'develop' into feature/CXSPA-574

justinlee01 committed:
Merge branch 'develop' into feature/CXSPA-574

plabadie commented on projects/core/src/occ/adapters/user/occ-user-cost-centers.adapter.ts:
```diff
@@ -25,8 +26,15 @@ export class OccUserCostCenterAdapter implements UserCostCenterAdapter {
   ) {}
 
   loadActiveList(userId: string): Observable<EntitiesModel<CostCenter>> {
+    // @see feature-libs/asm/root/interceptors/user-id.interceptor.ts
```
These comments are nice, but they come with a maintenance overhead.
Perhaps looking up the token references in the ide is enough for devs to understand the usage?

If we decide to keep them, lets update the path.

justinlee01 committed:
Moving comment from occ adapters to their proper place in the OccAsmContext file

justinlee01 committed:
Merge branch 'develop' into feature/CXSPA-574

justinlee01 committed:
Hiding UserIdHttpHeaderInterceptor feature behind a feature flag, as it requires an update to Commerce Cloud side. Also disabling the E2E tests written

justinlee01 commented:
@armanozak @plabadie As the feature requires a corresponding change to Commerce Cloud, I have hidden it behind a feature flag, implemented in "UserIdHttpHeaderInterceptor". I have also skipped the written e2e tests for the same reason. Let me know your thoughts.

justinlee01 committed:
Merge branch 'develop' into feature/CXSPA-574

justinlee01 committed:
Merge branch 'develop' into feature/CXSPA-574

justinlee01 committed:
Renaming OccAsmContext -> OccHttpContext and OCC_ASM_TOKEN -> OCC_HTTP_TOKEN, to make the context more generic

plabadie commented on projects/storefrontlib/feature-toggles.ts:
```diff
@@ -26,5 +26,16 @@ export interface FeatureToggles {
+     *
+     * Enabling this feature will send the emulated user's IDs on requests that need it as context.
+     */
+    enableCommerceCloudUserIdHeader?: boolean;
```
Good catch for the breaking change due to core filter settings.

I get the main reason for the feature flag is to prevent issues to customers when they upgrade to 5.1.  They would get errors form the requests that send the new header.
However it doesn't seem that this is a feature in of itself that customers want to toggle.

Do you think this behaviour could be hidden behind a 5.1 feature level flag instead?  Whenever a feature level flag is possible, we prefer that vs accumulating feature toggles.  (we have very few feature toggles.). Let's discuss.

In 6.0, it should become enabled by default.

justinlee01 replied:
Aren't there cases where the customer upgrades Spartacus but does not upgrade Commerce Cloud?
This type of toggle seems more pragmatic.

plabadie replied:
I'm not sure I understand.  Let's discuss.

plabadie commented on projects/storefrontapp-e2e-cypress/cypress/integration/regression/asm/asm.e2e-spec.ts:
```diff
@@ -72,4 +84,41 @@ context('Assisted Service Module', () => {
+    });
+
+    // This test only works if "sap-commerce-cloud-user-id" is added to the allowed headers of "corsfilter.commercewebservices.allowedHeaders" on the Commerce Cloud side
+    it.skip("should fetch products in a category based on the emulated user's authentication", () => {
```
We could update the CI backends to accept this header and enable the tests.
If we decide to go for a feature level flag, we might not have a choice but to update the CI backends before we merge this PR.  Let's discuss.

plabadie commented on feature-libs/asm/root/interceptors/user-id-http-header.interceptor.ts:
```diff
@@ -0,0 +1,83 @@
+      return next.handle(httpRequest);
+    }
+
+    const asmContext = httpRequest.context.get(OCC_HTTP_TOKEN);
```
perhaps rename `asmContext` to something like `occHttpContext`

justinlee01 replied:
Tbh I don't see a difference. In this case this context is still acting as an ASM context.

plabadie commented on projects/core/src/occ/utils/occ-http-token.ts:
```diff
@@ -0,0 +1,20 @@
+   * to avoid ambiguity.
+   * 'string' if the request has the user ID at hand.
+   * 'boolean' (or, 'true') if the request does not have the user ID but still needs to send the emulated ID.
+   * @see feature-libs/asm/root/interceptors/user-id-http-header.interceptor.ts
```
I think this comment is also generated and published online, in which case the file path won't mean much.  Perhaps we use the class name and import path to refer to other classes.  Not sure what we do elsewhere for these cases.

justinlee01 replied:
I can remove this line, but I'd like to see where this type of documentation gets published. This is an unusual obstacle for writing notes.
If the solution is to remove the JSDoc tag, that works for me.

plabadie replied:
It's [this doc](https://sap.github.io/spartacus/injectables/AsmService.html), where each element of the Public API is documented.

However, I just saw that _all_ the api elements in this doc are referenced from their source code file path, so forget about my objections.

plabadie commented on feature-libs/asm/root/interceptors/user-id-http-header.interceptor.ts:
```diff
@@ -0,0 +1,83 @@
+      concatMap((userId) => {
+        if (userId) {
+          const request = httpRequest.clone({
+            headers: httpRequest.headers.set(this.userIdHeader, userId),
```
There should be documentation updates that go with this PR.
We need to update the installation/configuration guides for the backend to mention the new allowed header that needs to be added in the config.

justinlee01 replied:
Draft here: https://github.com/SAP/spartacus-docs/pull/1492

justinlee01 commented:
As a note, I have created a ticket for updating the backend: https://jira.tools.sap/browse/CXSPA-1355
I myself am not sure how this is done - through updating `spartacussampledata` or other.

ercultimate commented on feature-libs/asm/root/interceptors/user-id-http-header.interceptor.spec.ts:
```diff
@@ -0,0 +1,226 @@
+import { Observable, of } from 'rxjs';
+import { take } from 'rxjs/operators';
+
+import { OCC_USER_ID_CURRENT } from '../../../../projects/core/src/occ/utils/occ-constants';
+import { UserIdService } from '../../../../projects/core/src/auth/user-auth/facade';
+import {
+  OCC_HTTP_TOKEN,
+  OCC_USER_ID_CONSTANTS,
+} from '../../../../projects/core/src/occ/utils';
```
Are these imports appropriate?

justinlee01 replied:
Ah, no. I suppose these are ignored for compilation because they're in unit tests, but they ought to be fixed. Ty.

justinlee01 committed:
Correcting imports in user-id-http-header.interceptor.spec.ts

justinlee01 committed:
Moving feature toggle for enabling sap-commerce-cloud-user-id header to AsmConfig, which is more module-appropriate

justinlee01 committed:
Merge branch 'develop' into feature/CXSPA-574

justinlee01 committed:
Copying AsmConfig and deprecating the old one, as the configuration should be in /root instead of /core

plabadie commented on feature-libs/asm/core/config/asm-config.ts:
```diff
@@ -11,6 +11,9 @@ import { Config, OccConfig } from '@spartacus/core';
   useExisting: Config,
 })
+/**
+ * @deprecated. Please use feature-libs/asm/root/asm-config.ts
```
The deprecated message should use the import path instead of the file path.  The import path is what the customers need to use the newer/non-deprecated class.

plabadie commented on feature-libs/asm/core/config/default-asm-config.ts:
```diff
@@ -6,6 +6,9 @@
 
+/**
+ * @deprecated. Please use feature-libs/asm/root/default-asm-config.ts
+ */
```
Was defaultAsmConfig part go the public API?

plabadie replied:
If not, perhaps we can move this one directly without deprecating.

justinlee01 replied:
Hmm. `AsmConfig` will live in `core`, but `defaultAsmConfig` will live in `root`. Let me see if the dependency framework is ok with that (as in, we `provideDefaultConfig` in root).

justinlee01 committed:
Moving defaultAsmConfig from core to root
Change import paths for deprecation message to be relative rather than absolute for AsmConfig

justinlee01 committed:
Removing implementation of core/config/asm-config.ts, instead extending it to use the implementation of root/config/asm-config.ts

justinlee01 committed:
Changing documentation for AsmConfig in @spartacus/asm/core as the config is not actually deprecated
Restoring ambient declaration in core config file
Changing userIdHttpHeaderInterceptor -> userIdHttpHeader
Slight modification of comment in user-id-http-header.interceptor.ts

justinlee01 committed:
Adding ticket number for currently skipped tests awaiting backend changes

justinlee01 committed:
Merge branch 'develop' into feature/CXSPA-574

justinlee01 committed:
Merge branch 'develop' into feature/CXSPA-574

justinlee01 committed:
Merge branch 'develop' into feature/CXSPA-574

justinlee01 committed:
Merge branch 'develop' into feature/CXSPA-574
